agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/utils/ResourceParser.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.azure.monitor.opentelemetry.exporter.implementation.utils;
+
+import com.azure.monitor.opentelemetry.exporter.implementation.builders.AbstractTelemetryBuilder;
+import com.azure.monitor.opentelemetry.exporter.implementation.models.ContextTagKeys;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+import java.util.Map;
+
+public final class ResourceParser {
+
+  public static void updateRoleNameAndInstance(
+      AbstractTelemetryBuilder builder, Resource resource) {
+    Map<String, String> existingTags = builder.build().getTags();
```

Comment:
this _looks_ expensive call `build` just to get the tags, though it's not really expensive since we have a slightly unusual builder here

I think my recommendation would be to remove the `System.getenv("HOSTNAME")` default from here, since that's what's causing the issue, and move that default into the standalone exporter's "telemetryInitializer"

Reply:
and then this class is just purely mapping otel `service.*` into ai `cloud.*`

Reply:
it still requires builder.build().getTags() check to make sure it doesn't overwrite customer's role name. it doesn't save much. besides hostname is part of the spec for roleinstance's default value.

Reply:
what if we change the "telemetry initializer" from `Consumer<AbstractTelemetryBuilder>` to `BiConsumer<AbstractTelemetryBuilder, Resource>`, and then we could call `ResourceParser` just from the standalone exporters and not affect the agent?

in the future we _might_ want to also map otel `service.*` attributes in the agent, but I think we will want to think carefully about order of precedence there

Reply:
changes are in standalone exporters though. agent is using the standalone exporter. i'm not sure not affect the agent part.

Reply:
> what if we change the "telemetry initializer" from `Consumer<AbstractTelemetryBuilder>` to `BiConsumer<AbstractTelemetryBuilder, Resource>`, and then we could call `ResourceParser` just from the standalone exporters and not affect the agent?

I sent a couple of commits to your branch to show what I meant, I think it turned out nice, also uncovered a couple of things we had missed previously